src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs

```diff
@@ -1599,14 +1627,43 @@ private void WindowClosing(object sender, CancelEventArgs e)
         }
 
         /// <summary>
-        /// Saves the state of currently displayed extension windows. This is needed because the closing event is
+        /// Saves the state of currently displayed extension windows and tabs. This is needed because the closing event is
         /// not called on child windows: https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing
         /// </summary>
-        private void SaveExtensionWindowsState()
+        private void SaveExtensionsState()
         {
+            //loop over all active extension windows and tabs.
             foreach (var window in ExtensionWindows.Values)
             {
                 SaveExtensionWindowSettings(window);
+                SaveExtensionOpenState(window);
+            }
+            //for any new extensions that are opened for the first time
+            foreach (var tab in ExtensionTabItems)
+            {
+                SaveExtensionOpenState(tab);
+            }
+            //update open state of all existing view extension in setting, if option to remember extensions is enabled in preferences
+            if (dynamoViewModel.PreferenceSettings.EnablePersistExtensions) {
+                var settings = dynamoViewModel.PreferenceSettings.ViewExtensionSettingsList;
+                foreach (var setting in settings)
+                {
+                    if (!ExtensionTabItems.Any(e => e.Uid == setting.UniqueId) && !ExtensionWindows.Values.Any(e => e.Uid == setting.UniqueId))
+                    {
+                        setting.IsOpen = false;
+                    }
+                }
+            }
+        }
+        //This method is to ensure that the extensions states are correctly saved for newly added extensions.
+        private void SaveExtensionOpenState(object o)
+        {
+            if (!dynamoViewModel.PreferenceSettings.EnablePersistExtensions || o == null) return;
+            var extId = o is TabItem ? ((TabItem)o).Uid : ((ExtensionWindow)o).Uid;
```

Comment:
this should probably be a type switch or other pattern match with a default case that fails gracefully with a message.

Reply:
message in logger?